### PR TITLE
[region-isolation] Fix join to be a true union operation including the symmetric difference.

### PR DESCRIPTION
**Pull request by gottesmm**
We were performing a union on the intersection of the lhs/rhs but were dropping
the parts of lhs/rhs that were in the symmetric difference of the two sets.

Without this, we would not diagnose cases like this where we had elements on the
lhs/rhs that were not in the intersection.

```
var closure: () -> () = {}

await transferToMain(closure)

if await booleanFlag {
   closure = {
     print(self.klass)
   }
} else {
   closure = {}
}
// At this point we would lose closure since they were different elements

await transferToMain(closure) // We wouldn't error on this!
```

rdar://117437059

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
I messed up a little bit of the output from one test for without region isolation when I updated it for region isolation... oops. Posting a commit.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: include/swift/SILOptimizer/Utils/PartitionUtils.h [hunk 1]
```diff
@@ -351,6 +356,21 @@ class Partition {
     return p;
   }
 
+  static Partition separateRegions(ArrayRef<Element> indices) {
+    Partition p;
+    if (indices.empty())
+      return p;
+
+    Region max_index =
+        Region(*std::max_element(indices.begin(), indices.end()));
```
Comment (xedin) [inline include/swift/SILOptimizer/Utils/PartitionUtils.h:365]: Maybe `fresh_label` could be assigned after we went over indices, seems like `max_index` could be computed by that loop?

File: unittests/SILOptimizer/PartitionUtilsTest.cpp [hunk 2]
```diff
@@ -100,6 +125,235 @@ TEST(PartitionUtilsTest, TestMergeAndJoin) {
...
+  ASSERT_EQ(tester.getRegion(35), 34);
+  ASSERT_EQ(tester.getRegion(36), 13);
+  ASSERT_EQ(tester.getRegion(37), 13);
+  ASSERT_EQ(tester.getRegion(38), 13);
+  ASSERT_EQ(tester.getRegion(39), 39);
+  ASSERT_EQ(tester.getRegion(40), 13);
+  ASSERT_EQ(tester.getRegion(41), 13);
+  ASSERT_EQ(tester.getRegion(42), 33);
+  ASSERT_EQ(tester.getRegion(43), 13);
+  ASSERT_EQ(tester.getRegion(44), 3);
```
Comment (xedin) [inline unittests/SILOptimizer/PartitionUtilsTest.cpp:354]: A suggestion for future testing - random number of elements assigned randomly, we could keep track of all of these assignments in a map and double-check that join is correct via double-checking that map in a loop.

**Review by xedin — APPROVED** [review]

File: unittests/SILOptimizer/PartitionUtilsTest.cpp (hunk 2 above)
Comment (gottesmm) [reply unittests/SILOptimizer/PartitionUtilsTest.cpp:354]: @xedin I did that here. I just generated it offline.

File: include/swift/SILOptimizer/Utils/PartitionUtils.h (hunk 1 above)
Comment (gottesmm) [reply include/swift/SILOptimizer/Utils/PartitionUtils.h:365]: Sure.

**Comment by gottesmm** [comment]
I will fix @xedin 's suggestions in a follow on.

**Pull request merged**